src/libsyntax/ast.rs

```diff
@@ -1746,13 +1730,21 @@ impl StructFieldKind {
     }
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
+pub enum VariantKind {
+    Dict,
```

Comment:
>What is Dict? Short for dictionary? we don't generally use that terminology in Rust.

I borrowed `Dict` (and the whole `VariantKind`) from `rustc::middle::ty::VariantKind`. I don't know a better name, so I tried at least to be consistent with some existing terminology.

>This enum is for struct kinds now, right? Not variant kinds.

Struct kinds and variant kinds are supposed to be the same thing.

>Apart from naming I find this enum a little bit weird - what kind of struct it is really just depends on the kind of the fields

We need to distinguish `{}`-structs, `();`-structs and `;`-structs even if they have no fields at all, because they add different things into value namespace (nothing, constructor function or constant, respectively). The "unit = tuple with no fields" approach used now leads to bugs in corner cases like https://github.com/rust-lang/rust/issues/28692 or https://github.com/rust-lang/rust/issues/16819#issuecomment-139509861